src/Navbar.js

```diff
@@ -122,103 +122,82 @@ const defaultProps = {
-
+const Navbar = React.forwardRef((props, ref) => {
+  let {
+    bsPrefix,
+    expand,
+    variant,
+    bg,
+    fixed,
+    sticky,
+    className,
+    children,
+    // Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
+    as: Component = 'nav',
+    expanded,
+    onToggle,
+    onSelect,
+    collapseOnSelect,
+    ...controlledProps
+  } = useUncontrolled(props, {
+    expanded: 'onToggle',
+  });
+
+  bsPrefix = useBootstrapPrefix(bsPrefix, 'navbar');
+
+  const handleCollapse = (...args) => {
     if (onSelect) onSelect(...args);
     if (collapseOnSelect && expanded) {
       onToggle(false);
     }
   };
 
-  handleToggle = () => {
-    const { onToggle, expanded } = this.props;
-
+  const handleToggle = () => {
     onToggle(!expanded);
   };
 
-  render() {
-    const {
-      bsPrefix,
-      expand,
-      variant,
-      bg,
-      fixed,
-      sticky,
-      className,
-      children,
-      // Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
-      as: Component = 'nav',
-      expanded: _1,
-      onToggle: _2,
-      onSelect: _3,
-      collapseOnSelect: _4,
-      ...props
-    } = this.props;
-
-    // will result in some false positives but that seems better
-    // than false negatives. strict `undefined` check allows explicit
-    // "nulling" of the role if the user really doesn't want one
-    if (props.role === undefined && Component !== 'nav') {
-      props.role = 'navigation';
-    }
-    let expandClass = `${bsPrefix}-expand`;
-    if (typeof expand === 'string') expandClass = `${expandClass}-${expand}`;
-
-    return (
-      <NavbarContext.Provider value={this.state.navbarContext}>
-        <SelectableContext.Provider value={this.handleCollapse}>
-          <Component
-            {...props}
-            className={classNames(
-              className,
-              bsPrefix,
-              expand && expandClass,
-              variant && `${bsPrefix}-${variant}`,
-              bg && `bg-${bg}`,
-              sticky && `sticky-${sticky}`,
-              fixed && `fixed-${fixed}`,
-            )}
-          >
-            {children}
-          </Component>
-        </SelectableContext.Provider>
-      </NavbarContext.Provider>
-    );
+  // will result in some false positives but that seems better
+  // than false negatives. strict `undefined` check allows explicit
+  // "nulling" of the role if the user really doesn't want one
+  if (controlledProps.role === undefined && Component !== 'nav') {
+    controlledProps.role = 'navigation';
   }
-}
+  let expandClass = `${bsPrefix}-expand`;
+  if (typeof expand === 'string') expandClass = `${expandClass}-${expand}`;
+
+  return (
+    <NavbarContext.Provider value={{ handleToggle, bsPrefix, expanded }}>
```

Comment:
we should `useMemo` the `value` here, otherwise React will trigger an update of any Consumers of this every time Navbar renders, since the object will not `===` the last object.